tools/2025-01-gitops-technical-design.md

```diff
@@ -0,0 +1,435 @@
+
+`deploy/images/bicep/Dockerfile`:
+```Dockerfile
+FROM alpine:latest
+
+ARG TARGETARCH
+
+RUN apk --no-cache add curl
+
+WORKDIR /
+
+RUN curl -Lo bicep https://github.com/Azure/bicep/releases/latest/download/bicep-linux-x64 \
+    && chmod +x ./bicep
+
+ENTRYPOINT ["/bin/sh"]
+```
+
+`deploy/Chart/templates/controller/deployment.yaml`:
+```yaml
+    spec:
+      ...
+      initContainers:
+      - name: bicep
+        image: "{{ .Values.bicep.image }}:{{ .Values.bicep.tag | default $appversion }}"
+        command: ['sh', '-c', 'mv ./bicep /usr/local/bin/bicep']
+        volumeMounts:
+        - name: bicep
+          mountPath: /usr/local/bin
+      ...
+      containers:
+      ...
+        volumeMounts:
+        - name: bicep
+          mountPath: /usr/local/bin
+      ...
+      volumes:
+      ...
+        - name: bicep
+          emptyDir: {}
+      ...
+```
+
+### Implementation Details
+<!--
+High level description of updates to each component. Provide information on
+the specific sub-components that will be updated, for example, controller, processor, renderer,
+recipe engine, driver, to name a few.
+-->
+
+- **Controller**: The Radius Flux Controller will be implemented as a new controller in the `controller` service. The controller will read the Git repository, compile Bicep files, and create DeploymentTemplate resources on the cluster.
+- **Bicep Compilation**: The Bicep compilation will be done using the Bicep CLI. The Bicep CLI itself will be present in a container image that is mounted as a volume in the controller pod. The controller will use the Bicep CLI to compile the Bicep file and return the compiled template as a string.
+
+### Error Handling
+<!--
+Describe the error scenarios that may occur and the corresponding recovery/error handling and user experience.
+-->
+
+All errors will be logged to the Kubernetes event log. We will also provide detailed error messages in the controller logs to help users troubleshoot any issues.
+
+## Test plan
+
+<!--
+Include the test plan to validate the features including the areas that
+need functional tests.
+
+Describe any functionality that will create new testing challenges:
+- New dependencies
+- External assets that tests need to access
+- Features that do I/O or change OS state and are thus hard to unit test
+-->
+
+- **Unit Tests**: Test the individual components of the Radius Flux Controller. There are many external dependencies that will need to be mocked, such as GitRepository objects and Bicep compilation.
+- **Functional Tests**: Test the end-to-end functionality of the Radius Flux Controller together with the DeploymentTemplate controller. This will involve creating GitRepository objects on a real Kubernetes cluster and verifying that the resources in the Bicep templates are created and updated correctly.
```

Comment:
Thanks for including test plan section. Do we need to think about any error scenarios and performance issues for large git repos?

Reply:
this is a good callout - In the controller code we can configure the `MaxDownloadSize` and `MaxUntarSize` both currently set to `Unlimited` by default. I can do some more testing on this